File: src/main/java/com/now/nowbot/service/MessageServiceImpl/GetNameService.java
package com.now.nowbot.service.MessageServiceImpl;

import com.now.nowbot.config.Permission;
import com.now.nowbot.qq.event.MessageEvent;
import com.now.nowbot.service.MessageService;
import com.now.nowbot.service.OsuApiService.OsuScoreApiService;
import com.now.nowbot.service.OsuApiService.OsuUserApiService;
import com.now.nowbot.throwable.GeneralTipsException;
import com.now.nowbot.util.DataUtil;
import com.now.nowbot.util.Instructions;
import jakarta.annotation.Resource;
import org.springframework.stereotype.Service;
import org.springframework.util.CollectionUtils;
import org.springframework.util.StringUtils;

import java.util.regex.Matcher;

@Service("GET_NAME")
public class GetNameService implements MessageService<Matcher> {
    @Resource
    OsuUserApiService userApiService;
    @Resource
    OsuScoreApiService scoreApiService;

    @Override
    public boolean isHandle(MessageEvent event, String messageText, DataValue<Matcher> data) throws Throwable {
        var m = Instructions.GET_NAME.matcher(messageText);
        if (m.find()) {
            data.setValue(m);
            return true;
        } else return false;
    }

    @Override
    public void HandleMessage(MessageEvent event, Matcher matcher) throws Throwable {
        var from = event.getSubject();

        if (! Permission.isGroupAdmin(event)) {
            throw new GeneralTipsException(GeneralTipsException.Type.G_Permission_Group);
        }

        var ids = DataUtil.splitString(matcher.group("data"));

        if (CollectionUtils.isEmpty(ids)) throw new GeneralTipsException(GeneralTipsException.Type.G_Fetch_List);

        StringBuilder sb = new StringBuilder();

        for (var i : ids) {
            if (! StringUtils.hasText(i)) {
                break;
            }

            long id;
            String name;

            try {
                id = Long.parseLong(i);
            } catch (NumberFormatException e) {
                sb.append("id=").append(i).append(" can't parse").append(',').append(' ');
                break;
            }

            try {
                name = userApiService.getPlayerOsuInfo(id).getUsername();
            } catch (Exception e) {
                sb.append("id=").append(id).append(" not found").append(',').append(' ');
                break;
            }

            sb.append(name).append(',').append(' ');
        }


        from.sendMessage(sb.substring(0, sb.length() - 2));
    }
}

File: src/main/java/com/now/nowbot/service/messageServiceImpl/GetNameService.kt
package com.now.nowbot.service.messageServiceImpl

import com.now.nowbot.dao.BindDao
import com.now.nowbot.entity.ServiceCallStatistic
import com.now.nowbot.model.osu.MicroUser
import com.now.nowbot.qq.event.MessageEvent
import com.now.nowbot.restrict.RestrictUtils.isCommonUser
import com.now.nowbot.service.MessageService
import com.now.nowbot.service.MessageService.DataValue
import com.now.nowbot.service.osuApiService.OsuUserApiService

import com.now.nowbot.throwable.botRuntimeException.IllegalStateException
import com.now.nowbot.throwable.botRuntimeException.PermissionException
import com.now.nowbot.util.DataUtil.splitString
import com.now.nowbot.util.Instruction
import org.springframework.stereotype.Service
import java.util.regex.Matcher

@Service("GET_NAME")
class GetNameService(
    private val userApiService: OsuUserApiService,
    private val bindDao: BindDao
) : MessageService<Matcher> {

    @Throws(Throwable::class) override fun isHandle(
        event: MessageEvent,
        messageText: String,
        data: DataValue<Matcher>
    ): Boolean {
        val m = Instruction.GET_NAME.matcher(messageText)
        if (m.find()) {
            data.value = m
            return true
        } else return false
    }

    @Throws(Throwable::class) override fun handleMessage(event: MessageEvent, param: Matcher): ServiceCallStatistic? {
        if (event.isCommonUser()) {
            throw PermissionException.DeniedException.BelowGroupAdministrator()
        }

        val sb = StringBuilder()

        // 使用批量获取
        val ids = if (event.hasAt()) {
            bindDao.getBindFromQQs(event.targets).map { it.userID }
        } else {
            splitString(param.group("data"), splitSpace = true)?.mapNotNull { it.toLongOrNull() }
        }

        if (ids.isNullOrEmpty()) throw IllegalStateException.Fetch("玩家编号")

        val nameMap: Map<Long, MicroUser> = userApiService.getMicroUsers(ids)
            .associateBy { it.userID }

        ids.forEach {
            val user = nameMap[it]

            sb.append(user?.username ?: "-").append(',')
        }

        event.replyAsync(sb.toString().removeSuffix(","))

        /*
        for (i in idStrs) {
            if (i.isBlank()) {
                continue
            }

            val id = try {
                i.toLong()
            } catch (e: NumberFormatException) {
                sb.append("id=").append(i).append(" can't parse").append(',')
                continue
            }

            val name = try {
                userApiService.getPlayerInfo(id).username
            } catch (e: Exception) {
                sb.append("id=").append(id).append(" not found").append(',')
                break
            }

            sb.append(name).append(',')
        }


        from.sendMessage(sb.substring(0, sb.length - 2))

         */


        return ServiceCallStatistic.builds(event, userIDs = ids)
    }
}
